chain/sync.go

```diff
@@ -573,15 +589,40 @@ func (syncer *Syncer) Sync(ctx context.Context, maybeHead *types.TipSet) error {
 		return nil
 	}
 
-	if err := syncer.collectChain(ctx, maybeHead, hts); err != nil {
+	ss := extractSyncState(ctx)
+	ss.Init(hts, maybeHead) // State `StageHeaders`.
+
+	subChain, err := syncer.collectChain(ctx, maybeHead, hts)
+	if err != nil {
 		span.AddAttributes(trace.StringAttribute("col_error", err.Error()))
 		span.SetStatus(trace.Status{
 			Code:    13,
 			Message: err.Error(),
 		})
+		ss.Error(err)
 		return xerrors.Errorf("collectChain failed: %w", err)
 	}
 
+	ss.SetStage(api.StageMessages)
+
+	// FIXME: `StageFetchingMessages` is buried inside this function (in
+	//  `iterFullTipsets`). We should extract it out and probably request
+	//  *both* block headers and messages together through the corresponding
+	//  chain exchange interface. We can do very little validation without the
+	//  messages so it shouldn't have much impact.
```

Comment:
Fetching messages has quite an impact on sync speed. We should not be doing it eagerly.
One of the most important parts of validation is not syncing headers we already failed to sync (and not fetching messages for them because we did not persist them).

Reply:
Fair point. We probably shouldn't request as much messages as chain we are fetching but it still seems to me we could benefit from eagerly requesting _some_ amount of messages in advance for upcoming blocks. We are currently at the other extreme of only requesting at the same moment of validation stalling the pipeline. Anyway, just a note, not going to change any of that here.